airbase/fetch.py

```diff
@@ -0,0 +1,274 @@
+    *,
+    timeout: float | None = None,
+    encoding: str | None = None,
+) -> list[dict[str, str]]:
+    """Request url and read response’s body as JSON
+
+    :param url: requested url
+    :param timeout: maximum time to complete request (seconds)
+    :param encoding: text encoding used for decodding the response's body
+
+    :return: decoded text from response's body as JSON
+    """
+    text = fetch_text(url, timeout=timeout, encoding=encoding)
+    payload: dict[str, str] | list[dict[str, str]]
+    payload = json.loads(text)
+    if isinstance(payload, dict):
+        return [payload]
+    return payload
+
+
+@overload
+def fetcher(
+    urls: list[str],
+    *,
+    encoding: str | None = None,
+    progress: bool = DEFAULT.progress,
+    raise_for_status: bool = DEFAULT.raise_for_status,
+    max_concurrent: int = DEFAULT.max_concurrent,
+) -> AsyncIterator[str]:
+    ...
+
+
+@overload
+def fetcher(
+    urls: dict[str, Path],
+    *,
+    encoding: str | None = None,
+    progress: bool = DEFAULT.progress,
+    raise_for_status: bool = DEFAULT.raise_for_status,
+    max_concurrent: int = DEFAULT.max_concurrent,
+) -> AsyncIterator[Path]:
+    ...
+
+
+async def fetcher(
+    urls: list[str] | dict[str, Path],
+    *,
+    encoding: str | None = None,
+    progress: bool = DEFAULT.progress,
+    raise_for_status: bool = DEFAULT.raise_for_status,
+    max_concurrent: int = DEFAULT.max_concurrent,
+) -> AsyncIterator[str | Path]:
+    """Request multiple urls and write resquest text into individual paths
+    it a `dict[url, path]` is provided, or return the decoded text from each request
+    if only a `list[url]` is provided.
+
+    :param urls: requested urls
+    :param encoding: text encoding used for decodding each response's body
+    :param progress: show progress bar
+    :param raise_for_status: Raise exceptions if download links
+        return "bad" HTTP status codes. If False,
+        a :py:func:`warnings.warn` will be issued instead.
+    :param max_concurrent: maximum concurrent requests
+
+    :return: url text or path to downloaded text, one by one as the requests are completed
+    """
+
+    async with aiohttp.ClientSession() as session:
+        semaphore = asyncio.Semaphore(max_concurrent)
+
+        async def fetch(url: str) -> str:
+            async with semaphore:
+                async with session.get(url, ssl=False) as r:
+                    r.raise_for_status()
+                    text: str = await r.text(encoding=encoding)
+                    return text
+
+        async def download(url: str, path: Path) -> Path:
+            text = await fetch(url)
+            async with aiofiles.open(str(path), mode="w") as f:
+                await f.write(text)
+            return path
+
+        jobs: list[Awaitable[str | Path]]
+        if isinstance(urls, dict):
+            jobs = [download(url, path) for url, path in urls.items()]
+        else:
+            jobs = [fetch(url) for url in urls]
```

Comment:
For my information: I see `mypy` complains if you remove `jobs: list[Awaitable[str | Path]]`. Why is that the case?

(I'm not super familiar with `mypy`)

Reply:
The internal functions `download` and `fetch` have different return types.
Without the line explicitly declaring the `job` type, `mypy` suspects an error.

In more detail:
- `download` and `fetch` are async functions. Each function call returns an `Awaitable` object, which needs to submitted/queued/scheduled to the async loop and then `await`-ed them in order to get the result from the function call.
- `download` returns a `Path`, so the on the `if` clause `jobs` is an `list[Awaitable[Path]]`.
- `fetch` returns the `url` that was part of the input, so the `else` clause `jobs` is an `list[Awaitable[str]]`
- The `type_a | type_b` notation is equivalent `Union[type_a, type_b]`

In principle, `jobs` could be annotated as `list[Awaitable[str]] | list[Awaitable[Path]]` ,but `mypy` did not like it. Therefore I went for `list[Awaitable[str | Path]]`, which is more general but captures the type well enough to distinguish between the objects that need to be awaited from the actual result from the function call.

Reply:
Ah okay so it just doesn't like multiple types potentially being assigned to the same variable, without explicitly declaring the intention to do so, I understand. Thanks!